// @bpmn-io/form-js/dist/index.cjs
'use strict';

var formJsViewer = require('@bpmn-io/form-js-viewer');
var formJsEditor = require('@bpmn-io/form-js-editor');
var formJsPlayground = require('@bpmn-io/form-js-playground');



Object.defineProperty(exports, 'Form', {
	enumerable: true,
	get: function () { return formJsViewer.Form; }
});
Object.defineProperty(exports, 'createForm', {
	enumerable: true,
	get: function () { return formJsViewer.createForm; }
});
Object.defineProperty(exports, 'getSchemaVariables', {
	enumerable: true,
	get: function () { return formJsViewer.getSchemaVariables; }
});
Object.defineProperty(exports, 'schemaVersion', {
	enumerable: true,
	get: function () { return formJsViewer.schemaVersion; }
});
Object.defineProperty(exports, 'FormEditor', {
	enumerable: true,
	get: function () { return formJsEditor.FormEditor; }
});
Object.defineProperty(exports, 'createFormEditor', {
	enumerable: true,
	get: function () { return formJsEditor.createFormEditor; }
});
Object.defineProperty(exports, 'FormPlayground', {
	enumerable: true,
	get: function () { return formJsPlayground.Playground; }
});
