// dmn-js-literal-expression/lib/features/powered-by/components/PoweredByOverlayComponent.js
import { createVNode, createComponentVNode, createTextVNode } from "inferno";
import { Component } from 'inferno';
import Logo from 'dmn-js-shared/lib/components/Logo';
export default class PoweredByOverlayComponent extends Component {
  constructor(props) {
    super(props);
    this.state = {
      show: false
    };
    this.onClick = this.onClick.bind(this);
    this.onShow = this.onShow.bind(this);
  }
  onClick() {
    this.setState({
      show: false
    });
  }
  onShow() {
    this.setState({
      show: true
    });
  }
  componentWillMount() {
    const eventBus = this._eventBus = this.context.injector.get('eventBus');
    eventBus.on('poweredBy.show', this.onShow);
  }
  componentWillUnmount() {
    this._eventBus.off('poweredBy.show', this.onShow);
  }
  render() {
    const show = this.state.show;
    return show && createVNode(1, "div", "powered-by-overlay", createVNode(1, "div", "powered-by-overlay-content", [createVNode(1, "a", "logo", createComponentVNode(2, Logo), 2, {
      "href": "https://bpmn.io",
      "target": "_blank",
      "rel": "noopener"
    }), createVNode(1, "span", null, [createTextVNode("Web-based tooling for BPMN, DMN and CMMN diagrams powered by "), createVNode(1, "a", null, createTextVNode("bpmn.io"), 2, {
      "href": "http://bpmn.io",
      "target": "_blank"
    }), createTextVNode(".")], 4)], 4, {
      "onClick": e => e.stopPropagation()
    }), 2, {
      "onClick": this.onClick
    });
  }
}
