// dmn-js-literal-expression/lib/features/textarea/components/TextareaEditorComponent.js
import { createVNode, createComponentVNode } from "inferno";
import { Component } from 'inferno';
import EditableComponent from 'dmn-js-shared/lib/components/EditableComponent';
export default class TextareaEditorComponent extends Component {
  constructor(props, context) {
    super(props, context);
    this._modeling = context.injector.get('modeling');
    this._viewer = context.injector.get('viewer');
    this.editLiteralExpressionText = this.editLiteralExpressionText.bind(this);
    this.onElementsChanged = this.onElementsChanged.bind(this);

    // there is only one single element
    const _this$getLiteralExpre = this.getLiteralExpression(),
      id = _this$getLiteralExpre.id;
    context.changeSupport.onElementsChanged(id, this.onElementsChanged);
  }
  getLiteralExpression() {
    return this._viewer.getDecision().decisionLogic;
  }
  onElementsChanged() {
    this.forceUpdate();
  }
  editLiteralExpressionText(text) {
    this._modeling.editLiteralExpressionText(text);
  }
  render() {
    // there is only one single element
    const _this$getLiteralExpre2 = this.getLiteralExpression(),
      text = _this$getLiteralExpre2.text;
    return createComponentVNode(2, Editor, {
      "className": "textarea editor",
      "value": text,
      "onChange": this.editLiteralExpressionText
    });
  }
}
class Editor extends EditableComponent {
  render() {
    return createVNode(1, "div", this.getClassName(), this.getEditor(), 0);
  }
}
