// dmn-js-decision-table/lib/features/simple-date-edit/components/OutputDateEdit.js
import { createVNode, createComponentVNode, createTextVNode } from "inferno";
import { Component } from 'inferno';
import ValidatedInput from 'dmn-js-shared/lib/components/ValidatedInput';
import { getSampleDate, validateISOString, parseString } from '../Utils';
export default class OutputDateEdit extends Component {
  constructor(props, context) {
    super(props, context);
    this._translate = context.injector.get('translate');
    this._modeling = context.injector.get('modeling');
    const element = this.props.context.element;
    const parsedString = parseString(element.businessObject.text);
    this.state = {
      date: parsedString ? parsedString.date : ''
    };
    const debounceInput = context.injector.get('debounceInput');
    this.debouncedEditCell = debounceInput(this.editCell.bind(this));
    this.editCell = this.editCell.bind(this);
    this.onClick = this.onClick.bind(this);
    this.onInput = this.onInput.bind(this);
  }
  editCell(cell, text) {
    this._modeling.editCell(cell, text);
  }
  onClick() {
    const element = this.props.context.element;
    const date = getSampleDate();
    this.setState({
      date
    });
    this.editCell(element.businessObject, "date(\"".concat(date, "\")"));
  }
  onInput(_ref) {
    let value = _ref.value;
    const element = this.props.context.element;
    this.setState({
      date: value
    });
    this.debouncedEditCell(element.businessObject, "date(\"".concat(value, "\")"));
  }
  render() {
    const date = this.state.date;
    return createVNode(1, "div", "context-menu-container simple-date-edit", [createVNode(1, "h3", "dms-heading", this._translate('Edit date'), 0), createVNode(1, "h4", "dms-heading", this._translate('Set date'), 0), createVNode(1, "div", null, [createComponentVNode(2, ValidatedInput, {
      "onInput": this.onInput,
      "placeholder": this._translate('e.g. { example } ', {
        example: getSampleDate()
      }),
      "validate": string => validateISOString(string) && this._translate(validateISOString(string)),
      "value": date,
      "className": "dms-block"
    }), createVNode(1, "p", "dms-hint", [this._translate('Set date'), createTextVNode(" "), createVNode(1, "button", "use-today", this._translate('to today'), 0, {
      "type": "button",
      "onClick": this.onClick
    }), createTextVNode(".")], 0)], 4)], 4);
  }
}
